### DISCUSSION TITLE
Keycloak setup for Single SPA angular Project

### DISCUSSION THREAD

**Discussion by Bouchnak-Maher** [keycloak/keycloak#28958] [Q&A]
Hello everyone!

I'm in the process of developing an application using Single SPA. I've reached the point where I need to integrate Keycloak authentication into my micro frontend architecture.

My question is: Should I integrate Keycloak directly into the Single SPA root configuration or into all the other parcels, and what configurations should I add?

I'm attempting to integrate Keycloak just into the root config, but I'm encountering an issue where the requests coming from the parcels don't include the access token. Additionally, the parcels don't verify the expiration of the token or session.

Thank you!

**Comment by codespearhead** [comment]
You should use OAuth 2.x Authorization Code Flow with PKCE for public clients such as SPAs, as shown [in this React tutorial](https://medium.hexadefence.com/securing-a-react-app-using-keycloak-ac0ee5dd4bfc).

Keycloak's first-party [OIDC-compliant, client-side JavaScript client](https://www.npmjs.com/package/keycloak-js) should be treated as any other middleware, so the way to enforce a middleware in an application depends entirely on whatever framework you're using. Thus, it's out of scope of Keycloak and should be asked in the respective community.

I don't know Angular, but item 4.1 from [this tutorial](https://www.angulararchitects.io/en/blog/oauth2-with-spring-angular-keycloak-spring-for-resource-server/) might help.

**Comment by PeS82** [comment]
We are using 

https://www.npmjs.com/package/keycloak-angular

and

https://www.npmjs.com/package/angular-auth-oidc-client

They both work well, As for KC config, as was said, configure OIDC client that uses PKCE and you are good to go.
